### PR TITLE
add multi-field scoping for slugs

### PR DESCRIPTION
**Pull request by mikekosulin**
* Add option for multi-field scoping in slugs, expanding beyond single-field/relation scopes without altering existing functionality

### CODE REVIEW

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:3]: We haven't released 7.1.0. Let's increment the version to 7.1.0 in this PR, it should say `7.1.0 (Next)`.

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:5]: Make this like the rest of the lines with your GitHub username.

File: lib/mongoid/slug.rb
Comment (dblock) [inline lib/mongoid/slug.rb:95]: Probably simpler to do `Array(slug_scope).each do |...|`, avoids code duplication.

File: lib/mongoid/slug.rb
Comment (dblock) [inline lib/mongoid/slug.rb:134]: Same as above, `Array(slug_scope).each`

File: lib/mongoid/slug/unique_slug.rb
Comment (dblock) [inline lib/mongoid/slug/unique_slug.rb:105]: `Array(slug_scope).each ...`

**Review by dblock — COMMENTED** [review]
Looks great! Let's simplify the code a bit with a uniform `Array(...).each ...`

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:1]: Increment this in https://github.com/mongoid/mongoid-slug/blob/master/lib/mongoid/slug/version.rb as well please.

File: README.md
Comment (dblock) [inline README.md:214]: Nit: lowercase 'This' :)

File: lib/mongoid/slug.rb
Comment (dblock) [inline lib/mongoid/slug.rb:95]: The implementation of `slug_scope_key` iterates over slug scopes that can be `nil`, then returns `nil` in the case of `slug_scope` being nil.  

I suggest adding a `slug_scopes` method that does not call `slug_scope_key`:

```ruby
def slug_scopes
    slug_scope ? Array(slug_scope) : [nil]
end
```

File: lib/mongoid/slug/unique_slug.rb
Comment (dblock) [inline lib/mongoid/slug/unique_slug.rb:103]: Is the `scope` assignment unnecessary here? 

```
Array(slug_scope).each ... 
```

**Review by dblock — CHANGES_REQUESTED** [review]
Much better, but the implementation is still a little confusing. See below. Thanks for hanging in here with me!

File: lib/mongoid/slug.rb
Comment (mikekosulin) [reply lib/mongoid/slug.rb:95]: Got it.

Also I believe there's no need to change the way we invoke the index builder and we should keep it as it was in version 7.

```ruby
if slug_index && !embedded?
  Mongoid::Slug::IndexBuilder.build_indexes(self, slug_scope_key, slug_by_model_type,
                                            options[:localize])
end
```

But we've to modify the [index_builder.rb](https://github.com/mongoid/mongoid-slug/blob/6b6ad2f07f4772f5615dc8492c3ba870afb017c5/lib/mongoid/slug/index_builder.rb#L31) this way:

```diff
- fields[scope_key] = 1 if scope_key
+ Array(scope_key).each { |key| fields[key] = 1 }
```

Comment (dblock) [reply lib/mongoid/slug.rb]: As long as all test pass ;)

File: lib/mongoid/slug.rb
Comment (dblock) [inline lib/mongoid/slug.rb:125]: Should this be renamed to `slug_scope_keys`?

**Review by dblock — COMMENTED** [review]
I think version.rb is the last thing.

**Comment by dblock** [comment]
@johnnyshields any comments/issues?

File: lib/mongoid/slug.rb
Comment (mikekosulin) [reply lib/mongoid/slug.rb:125]: Makes sense

**Review by dblock — APPROVED** [review]
LGTM! Thanks. I'll wait for @johnnyshields to chime in, will merge this week if we don't hear from him.

**Comment by mikekosulin** [comment]
Hey @dblock, @johnnyshields any updates on merging? 😊

File: README.md
Comment (johnnyshields) [inline README.md:214]: Hmmm... shouldn't this still create only one _compound_ index? For example:

```
slug :name, scope: %i[company_id department_id]

# Should create
index company_id: 1, department_id: 1, _slugs: 1

# NOT
index company_id: 1, _slugs: 1
index department_id: 1, _slugs: 1
```

Comment (mikekosulin) [reply README.md]: Sorry for the late response. You're right, it creates only a compound index in this case. I've [fixed it here](https://github.com/mongoid/mongoid-slug/pull/274/commits/601318844bdaf81212754d4212868fb28aca8208) and forgot to remove the note.

File: lib/mongoid/slug.rb
Comment (johnnyshields) [inline lib/mongoid/slug.rb:310]: This was an unrelated rubocop failure

**Comment by johnnyshields** [comment]
@mikekosulin great work, merged! Thanks.

**Pull request merged**
